### PR TITLE
Adding ability to download KBs when running Update-DbaInstance

### LINKED ISSUE #5043: Option to download updates when running Update-DbaInstance
---
name: Option to download updates when running Update-DbaInstance.
---

### Summary of new feature

* ability to download updates from microsoft update catalog instead of relying on pre-downloaded updates in local folders.

### Proposed technical details (if applicable)

* reference: can use this idea https://github.com/exchange12rocks/WU/tree/master/Get-WUFilebyID to get the direct link and download

### Latest version of dbatools as of writing

* 0.9.752

### PR DESCRIPTION
**Pull request by nvarscar**
## Type of Change

 - [ ] Bug fix (non-breaking change, fixes  )
 - [x] New feature (non-breaking change, adds functionality, fixes #5043 )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [x] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [x] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
Add -Download functionality to Update-DbaInstance

### Approach
When enabled, function would download each needed KB to the temp/network drive and (if needed) copy it to the remote machine.

### Commands to test
Added an example showcasing the parameter usage.

### Learning
Start-BitsTransfer was annoying and didn't want to play nicely with PSDrive drives. Had to switch back to Copy-Item.

### CODE REVIEW

**Comment by nvarscar** [comment]
> It Update-DbaInstance.ps1 is not compliant with the OTBS formatting style. Please run Invoke-DbatoolsFormatter against the failing file and commit the changes.

I did, what does it want from me huh

**Comment by nvarscar** [comment]
I don't understand. The tests pass locally. Invoke-DbatoolsFormatter leaves the file as is. 
```
 $content = Get-Content -Raw ..\functions\Update-DbaInstance.ps1 -Encoding UTF8
 $result = Invoke-Formatter -ScriptDefinition $content -Settings CodeFormattingOTBS
 $result -eq $content
True
```

**Comment by potatoqualitee** [comment]
@niphlod can you take a look? this issue also occurred on another Draft PR and I couldn't resolve it.

**Comment by potatoqualitee** [comment]
k, i committed the code from dev which should have the fix 👍

**Comment by nvarscar** [comment]
Thanks for making it work, somehow @potatoqualitee ! It seems however that the fix broke the indentation - it's not what it's supposed to be now, as confirmed by 
```
}
}
```
at the end

**Comment by potatoqualitee** [comment]
mannnn how on earth does that happen? i made some changes that were suggested by vscode or whatever. reverted changes. can you help @niphlod ? D:

**Comment by niphlod** [comment]
whoops, I didn't receive the notification for this ! totally gonna inspect in a few hours

**Comment by wsmelton** [comment]
Should be fixed now.

**Comment by potatoqualitee** [comment]
That seems to have worked, @wsmelton ! It didn't break formatting at the bottom either, like mine

**Comment by potatoqualitee** [comment]
nope, i was wrong. gosh this is a miserable experience 

image

**Comment by wsmelton** [comment]
That can be ignored I think, and is likely coming from the script block. It is something to do with PSSA rule in indentation.

**Comment by potatoqualitee** [comment]
@nvarscar can you download the latest change and test in yours?

**Comment by wsmelton** [comment]
My version shows no indentation issues. If you look at the problems panel on yours @potatoqualitee it will show you the rules it thinks are violated.

**Comment by nvarscar** [comment]
It doesn't change the formatting on my end after @wsmelton 's commit. Thanks!

File: functions/Update-DbaInstance.ps1 [hunk 1]
```diff
@@ -289,13 +300,96 @@ function Update-DbaInstance {
...
+            if (([DbaInstanceParameter]$groupItem.ComputerName).IsLocalHost) {
+                $remoteFolder = $Destination
+            } else {
+                $uncFileName = Join-AdminUnc -ComputerName $ComputerName -Path $Destination
+                $driveSplat = @{
+                    Name       = 'UpdateCopy'
+                    Root       = $uncFileName
+                    PSProvider = 'FileSystem'
+                }
+                if ($Credential) { $driveSplat.Credential = $Credential }
```
Comment (wsmelton) [inline functions/Update-DbaInstance.ps1:358]: This is not required if you are going to splat the command. It handles dealing with `$Credential` not being provided.

File: functions/Update-DbaInstance.ps1 [hunk 2]
```diff
@@ -289,13 +300,96 @@ function Update-DbaInstance {
...
     process {
         if (Test-FunctionInterrupt) { return }
 
         #Resolve all the provided names
         $resolvedComputers = @()
-        $pathIsNetwork = $Path | ForEach-Object -Begin { $o = @() } -Process { $o += $_ -like '\\*' } -End { $o -contains $true }
+        $pathIsNetwork = $Path | Test-NetworkPath
         foreach ($computer in $ComputerName) {
             $null = Test-ElevationRequirement -ComputerName $computer -Continue
             if (!$computer.IsLocalHost -and -not $notifiedCredentials -and -not $Credential -and $pathIsNetwork) {
```
Comment (wsmelton) [inline functions/Update-DbaInstance.ps1:395]: If you would not mind also updating the `!` to be `-not` as well for consistency 😉

**Review by wsmelton — APPROVED** [review]
I don't have a system to test it on but it all looks good to me short a few items just noticed.

File: functions/Update-DbaInstance.ps1 (hunk 1 above)
Comment (nvarscar) [reply functions/Update-DbaInstance.ps1]: Sounds good, moved Credential to the splat. Didn't expect a native cmdlet to handle $null credentials properly.

File: functions/Update-DbaInstance.ps1 (hunk 2 above)
Comment (nvarscar) [reply functions/Update-DbaInstance.ps1]: fixed

**Comment by potatoqualitee** [comment]
oh yay, than you both! 💯 🚀 !

**Comment by garethnewman** [comment]
love it!!

**Pull request merged**